docs/contributors/code/coding-guidelines.md

```diff
@@ -136,10 +136,278 @@ export { __unstableDoTerribleAwfulAction } from './api';
+
+If your experimental API is only meant for the Gutenberg Plugin but not for the next WordPress major release, consider limiting the export to the plugin environment. For example, `@wordpress/components` could do that to receive early feedback about a new Component, but avoid bringing that component to WordPress core:
+
+```js
+if ( IS_GUTENBERG_PLUGIN ) {
+	export { __experimentalFunction } from './experiments';
+}
+```
+
+#### Replace experimental selectors with hooks
+
+In React components, you can replace public experimental selectors with non-exported React hooks:
+
+```js
+// Instead of this:
+	// selectors.js:
+	export function __unstableHasActiveBlockOverlayActive( state, parent ) { /* ... */ }
+	export function __unstableIsWithinBlockOverlay( state, clientId ) {
+		let parent = state.blocks.parents[ clientId ];
+		while ( !! parent ) {
+			if ( __unstableHasActiveBlockOverlayActive( state, parent ) ) {
+				return true;
+			}
+			parent = state.blocks.parents[ parent ];
+		}
+		return false;
+	}
+	// MyComponent.js:
+	function MyComponent({ clientId }) {
+		const { __unstableIsWithinBlockOverlay } = useSelect( myStore );
+		const isWithinBlockOverlay = __unstableIsWithinBlockOverlay( clientId );
+		// ...
+	}
+
+// Consider this:
+	// MyComponent.js:
+	function hasActiveBlockOverlayActive ( selectors, parent ) { /* ... */ }
+	function useIsWithinBlockOverlay( clientId ) {
+		return useSelect( ( select ) => {
+			const selectors = select( blockEditorStore ); 
+			let parent = selectors.getBlockRootClientId( clientId );
+			while ( !!parent ) {
+				if ( hasActiveBlockOverlayActive( selectors, parent ) ) {
+					return true;
+				}
+				parent = selectors.getBlockRootClientId( parent );
+			}
+			return false;
+		});
+	}
+	function MyComponent({ clientId }) {
+		const isWithinBlockOverlay = useIsWithinBlockOverlay( clientId );
+		// ...
+	}
+```
+
+If your experimental selector needs to access a part of the store that isn't
+accessible by the existing public selectors, you may access the store data directly
+via `useRegistry()`:
+
+```js
+function MyComponent() {
+	const registry = useRegistry();
+	const enabledFeatures = registry.stores['my-store'].store.getState().features;
+	// ...
+}
+```
+
+#### Replace experimental actions with hooks and thunks
+
+Public experimental actions can be replaced with non-exported React hooks:
+
+```js
+// Instead of this:
+	// actions.js:
+	export function __experimentalBeforeSave() {
+		return { type: 'BEFORE_SAVE' };
+	}
+	// MyComponent.js:
+	function MyComponent() {
+		const { __experimentalBeforeSave } = useDispatch( myStore );
+		const onClick = () => __experimentalBeforeSave();
+		// ...
+	}
+
+// Consider this:
+	// MyComponent.js:
+	function MyComponent() {
+		const registry = useRegistry();
+		const onClick = () => registry.stores['my-store'].store.dispatch({ type: 'BEFORE_SAVE' });
```

Comment:
Actually, this is a bad approach, we should always be using useDispatch to access actions. This stores key has always been meant as an internal variable. I'd remove this example from here.

Reply:
The only reason this "stores" key exist is to make other public APIs work like "use" (which I believe we should deprecate)

Reply:
Ah thank you, let's remove it indeed then

Reply:
> I don't think it solves the problem for selectors, actions (...)

@youknowriad note this doc fragment discusses _replacing experimental actions with hooks and thunks_. I'll remove it, but it does not affect the section about `lock()` and `unlock()` below:

```
unlock( select( store ) ).__experimentalHasContentRoleAttribute()
```